test-integration/routes.test.js

```diff
@@ -1,89 +1,97 @@
+import cloneDeep from "lodash.clonedeep";
 import { testAdminApi, tearDown } from "./util";
 import execute from "../src/core";
 
-const routeId = "c5b6914d-2bfd-4bb5-8f45-2e94cf9ff8b5";
-const path = "/foo";
-const endpoint = {
-    name: "route",
-    params: { routeId: routeId }
+const baseConfig = {
+    services: [
+        {
+            name: "example",
+            attributes: {
+                url: "http://example.com"
+            },
+            routes: [
+                {
+                    name: "foo",
+                    attributes: {
+                        paths: ["/foo", "/bar"],
+                        regex_priority: 3
+                    }
+                }
+            ]
+        }
+    ]
 };
 
-async function requestRoute(params) {
-    const response = await testAdminApi.requestEndpoint(endpoint, params);
-    return await response.json();
+// Assumes there's only one route configured
+async function getRoute() {
+    return (await testAdminApi.fetchRoutes())[0];
 }
 
-beforeEach(async () => {
-    await tearDown();
-    const response = await testAdminApi.requestEndpoint(
-        {
-            name: "services"
-        },
-        {
-            method: "POST",
-            body: {
-                name: "example",
-                url: "http://example.com"
-            }
-        }
+beforeEach(tearDown);
+
+it("throws an error if the route doesn't have a name", async () => {
+    const config = cloneDeep(baseConfig);
+    delete config.services[0].routes[0].name;
+    await expect(execute(config, testAdminApi)).rejects.toThrow(
+        /^Route name is required/
     );
-    const serviceId = (await response.json()).id;
-    await testAdminApi.requestEndpoint(endpoint, {
-        method: "PUT",
-        body: {
-            paths: [path],
-            service: {
-                id: serviceId
-            }
-        }
-    });
 });
 
-it("should update a route name when it's a top level property", async () => {
-    const config = {
-        services: [
-            {
-                name: "example",
-                attributes: {
-                    url: "http://example.com"
-                },
-                routes: [
-                    {
-                        name: "foo",
-                        id: routeId,
-                        attributes: {
-                            paths: [path]
-                        }
-                    }
-                ]
-            }
-        ]
-    };
+
+it('creates, updates, and removes a route', async () => {
+    await execute(baseConfig, testAdminApi);
+    expect(await getRoute()).toMatchObject({
+        name: "foo",
+        paths: expect.arrayContaining(["/foo", "/bar"]),
+        regex_priority: 3
+    });
```

Comment:
These integration tests are a little different from the ones we already have. Rather than snapshot testing the logs, I'm just asserting various things about the state of kong, which gives me more confidence in these changes.

Reply:
If this approach works for us, we'll probably end up wanting some utilities to make these tests easier to write.